packages/sources/coinpaprika-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,219 @@
+      percent_change_7d: number
+      percent_change_30d: number
+      percent_change_1y: number
+      ath_price: number
+      ath_date: string
+      percent_from_price_ath: number
+    }
+  }
+  cost?: number
+}
+
+export interface CryptoRequestParams {
+  coinid?: string
+  base?: string
+  quote: string
+}
+
+export type EndpointTypes = {
+  Request: {
+    Params: CryptoRequestParams
+  }
+  Response: {
+    Data: number
+    Result: number
+  }
+  CustomSettings: SettingsMap
+  Provider: {
+    RequestBody: CryptoRequestBody
+    ResponseBody: CryptoResponseSchema[]
+  }
+}
+
+export type EndpointTypesSingle = {
+  Request: {
+    Params: CryptoRequestParams
+  }
+  Response: {
+    Data: {
+      result: number
+    }
+    Result: number
+  }
+  CustomSettings: SettingsMap
+  Provider: {
+    RequestBody: CryptoRequestBody
+    ResponseBody: CryptoResponseSchema
+  }
+}
+
+interface ResultEntry {
+  value: number
+  params: {
+    quote: string
+    base?: string
+    coinid?: string
+  }
+}
+
+const charsToEncode = {
+  ':': '%3A',
+  '/': '%2F',
+  '?': '%3F',
+  '#': '%23',
+  '[': '%5B',
+  ']': '%5D',
+  '@': '%40',
+  '!': '%21',
+  $: '%24',
+  '&': '%26',
+  "'": '%27',
+  '(': '%28',
+  ')': '%29',
+  '*': '%2A',
+  '+': '%2B',
+  ',': '%2C',
+  ';': '%3B',
+  '=': '%3D',
+  '%': '%25',
+  ' ': '%20',
+  '"': '%22',
+  '<': '%3C',
+  '>': '%3E',
+  '{': '%7B',
+  '}': '%7D',
+  '|': '%7C',
+  '^': '%5E',
+  '`': '%60',
+  '\\': '%5C',
+}
+
+const stringHasWhitelist = (str: string, whitelist: string[]): boolean =>
+  whitelist.some((el) => str.includes(el))
+
+const percentEncodeString = (str: string, whitelist: string[]): string =>
+  str
+    .split('')
+    .map((char) => {
+      const encodedValue = charsToEncode[char as keyof typeof charsToEncode]
+      return encodedValue && !whitelist.includes(char) ? encodedValue : char
+    })
+    .join('')
+
+export const buildUrlPath = (pathTemplate = '', params = {}, whitelist = ''): string => {
+  const allowedChars = whitelist.split('')
+
+  for (const param in params) {
+    const value = params[param as keyof typeof params]
+    if (!value) continue
+
+    // If string contains a whitelisted character: manually replace any non-whitelisted characters with percent encoded values. Otherwise, encode the string as usual.
+    const encodedValue = stringHasWhitelist(value, allowedChars)
+      ? percentEncodeString(value, allowedChars)
+      : encodeURIComponent(value)
+
+    pathTemplate = pathTemplate.replace(`:${param}`, encodedValue)
+  }
+
+  return pathTemplate
+}
+
+export const buildBatchedRequestBody = (
+  params: CryptoRequestParams[],
+  config: AdapterConfig,
+): HttpRequestConfig<CryptoRequestBody> => {
+  const headers: { Authorization?: string } = {}
+  if (config.API_KEY) {
+    headers['Authorization'] = config.API_KEY
+  }
+  return {
+    baseURL: config.API_KEY ? PRO_API_ENDPOINT : DEFAULT_API_ENDPOINT,
+    url: 'v1/tickers',
```

Comment:
The coinpaprika tickers response returns information for all active tickers on their end (many thousands). We should reevaluate this as it's been a problem before, even if batching is theoretically possible

Reply:
Since we can only batch 3 unique quotes in a single request, I noticed testing locally that the background execute takes longer to complete than the default cache max age (1.5min) when we have more than 6 unique quotes in the subscription set. In that scenario, we're calling the `v1/tickers` endpoint more than twice. Not sure what the alternative is though. The only other endpoint we can use is for a single base. @boxhock if you have any thoughts/suggestions.

Reply:
> I noticed testing locally that the background execute takes longer to complete than the default cache max age (1.5min) when we have more than 6 unique quotes in the subscription set

This is if you're on the lowest API rate limit tier, right? If so, I'd argue that it's simply a limitation of how many different requests you can send to Coingecko on the free tier.

Is this also happening on the paid tier?

Reply:
I was testing with a paid tier but realized I didn't set the tier in the env vars. That's probably what caused the slow background execute. I'll test again with that

Reply:
Retested with the rate limit tier set and it took 13 unique quotes without any issue. And that's not any sort of limit, it's just where I stopped testing. I'm ok with resolving this conversation if you guys are.

Reply:
I'm going to mark this as resolved since the v3 version seems to be capable of handling multiple calls to this endpoint per background execute cycle.